WordPress/gutenberg
# Add a warning when Cloudflare blocks REST API requests

## Description

When Cloudflare blocks REST API requests, there's nothing in the UI to suggest what the cause is, which is a confusing and frustrating experience.

Adding extra information when we detect the error helps folks determine their best course of action.

This is a temporary measure, to help with issues like #2704, until Cloudflare roll out their fix. As such, the message box is a fairly lazy copy/paste of the existing block warning in the Classic Editor.

## How has this been tested?

With Cloudflare:
- Enable the WAF on your site. You may need to disable rules `WP0003` and `WP0004`, to ensure you're not being whitelisted.
- Try to save a post.
- After the save fails, disable rules `WP0025A` and `WP0025B`.
- Try to save it again: the save should succeed.

On your local site, you can simulate the REST API being blocked by Cloudflare with this snippet:

```php
add_action( 'rest_api_init', function() {
	header( 'Content-Type: text/html' );
	status_header( 403 );
	echo '<span>Cloudflare Ray ID: foo</span>';
	die();
} );
```

## Screenshots

<img width="1279" alt="screen shot of the Cloudflare error message" src="https://user-images.githubusercontent.com/352291/43752933-a78c22d4-9a46-11e8-9d7b-15b53a18c629.png">

Clicking **Learn More** will redirect you to the Classic Editor, showing this message:

<img width="553" alt="Cloudflare detailed message" src="https://user-images.githubusercontent.com/352291/43752953-c369fa12-9a46-11e8-9046-eb2d1a43c147.png">

## Checklist:
